cds/modules/records/permissions.py

```diff
@@ -0,0 +1,184 @@
+    """Record permission.
+
+    - Create action given to any authenticated user.
+    - Read access given to everyone.
+    - Update access given to record owners.
+    - Delete access given to admins only.
+    """
+
+    create_actions = ['create']
+    read_actions = ['read']
+    read_files_actions = ['read-files']
+    update_actions = ['update']
+    delete_actions = ['delete']
+
+    def __init__(self, record, func, user):
+        """Initialize a file permission object."""
+        self.record = record
+        self.func = func
+        self.user = user or current_user
+
+    def can(self):
+        """Determine access."""
+        return self.func(self.user, self.record)
+
+    @classmethod
+    def create(cls, record, action, user=None):
+        """Create a record permission."""
+        if action in cls.create_actions:
+            return cls(record, allow, user)
+        elif action in cls.read_actions:
+            return cls(record, allow, user)
+        elif action in cls.read_files_actions:
+            return cls(record, has_read_files_permission, user)
+        elif action in cls.update_actions:
+            return cls(record, has_update_permission, user)
+        elif action in cls.delete_actions:
+            return cls(record, has_admin_permission, user)
+        else:
+            return cls(record, deny, user)
+
+
+class DepositPermission(RecordPermission):
+    """Deposit permission.
+
+    - Read action given to record owners.
+    - Delete action given to record owners (still subject to being unpublished)
+    """
+
+    @classmethod
+    def create(cls, record, action, user=None):
+        """Create a deposit permission."""
+        if action in cls.read_actions:
+            return cls(record, has_update_permission, user)
+        elif action in cls.delete_actions:
+            return cls(record, has_update_permission, user)
+        return super(DepositPermission, cls).create(record, action, user=user)
+
+
+#
+# Utility functions
+#
+def _get_user_provides():
+    """Extract the user's provides from g."""
+    return [str(need.value) for need in g.identity.provides]
+
+
+def deny(user, record):
+    """Deny access."""
+    return False
+
+
+def allow(user, record):
+    """Allow access."""
+    return True
+
+
+def has_read_files_permission(user, record):
+    """Check if user has read access to the record."""
+    # Allow everyone for now
+    # TODO: decide on files access rights
+    return True
+
+
+def has_update_permission(user, record):
+    """Check if user has update access to the record."""
+    # Allow owners
+    user_id = int(user.get_id()) if user.is_authenticated else None
+    if user_id in record.get('owners', []):
+        return True
+    deposit_owners = record.get('_deposit', {}).get('owners', [])
+    if user_id in deposit_owners:
+        return True
+    # Allow e-group members
+    user_groups = _get_user_provides()
+    if set(user_groups).intersection(set(deposit_owners)):
```

Comment:
`not set(user_groups).isdisjoint(set(deposit_owners))` is faster, as you don't need the actual elements but just to prove the existence of a single common element.